skbio/diversity/_block.py

```diff
@@ -0,0 +1,280 @@
+        if id_pairs:
+            kw = kwargs.copy()
+            kw['id_pairs'] = id_pairs
+            kw['row_ids'] = row_ids
+            kw['col_ids'] = col_ids
+
+            yield kw
+
+
+def _block_compute(**kwargs):
+    """Compute a block within the resulting distance matrix
+
+    Notes
+    -----
+    This method encapsulates the two expensive operations to perform for each
+    block, namely, the "shearing" of the phylogenetic tree to correspond to
+    only the OTUs of interest, and the actual beta diversity calculations.
+
+    Returns
+    -------
+    DistanceMatrix
+    """
+    block_kw = _block_party(**kwargs)
+
+    return partial_beta_diversity(**block_kw)
+
+
+def _map(func, kw_gen):
+    """Map a function over arguments
+
+    Note
+    ----
+    builtin map does not allow for mapping with kwargs.
+
+    """
+    for kwargs in kw_gen:
+        yield func(**kwargs)
+
+
+def _reduce(blocks):
+    """Reduce an iterable of partial distance matrices into a full matrix"""
+    all_blocks = list(blocks)
+
+    # Determine the maximum integer ID observed in the blocks. There exists a
+    # 1-1 mapping between the integer ID and a sample ID. We increment by 1
+    # as the integer ID space begins with zero, and we'll be using this value
+    # to determine the size of the resulting full distance matrix.
+    n_ids = max(map(lambda x: max(x.ids), all_blocks)) + 1
+
+    mat = np.zeros((n_ids, n_ids), dtype=float)
+
+    # TODO: something smarter.
+    for block in all_blocks:
+        n_blk_ids = len(block.ids)
+
+        # get the corresponding coordinates in the master matrix
+        master_idx = [(i, j) for row, i in enumerate(block.ids)
+                      for j in block.ids[row+1:]]
+
+        # get the corresponding coordinates within the current block
+        block_idx = [(i, j) for row, i in enumerate(range(n_blk_ids))
+                     for j in range(row+1, n_blk_ids)]
+
+        for (m_i, m_j), (b_i, b_j) in zip(master_idx, block_idx):
+            mat[m_i, m_j] += block.data[b_i, b_j]
+
+    return DistanceMatrix(mat + mat.T, list(range(n_ids)))
+
+
+@experimental(as_of="0.5.0-dev")
+def block_beta_diversity(metric, counts, ids, validate=True, k=64,
+                         reduce_f=_reduce, map_f=_map, **kwargs):
+    """Perform a block-decomposition beta diversity calculation
+
+    Parameters
+    ----------
+    metric : str or callable
+        The pairwise distance function to apply. If ``metric`` is a string, it
+        must be resolvable by scikit-bio (e.g., UniFrac methods), or must be
+        callable.
+    counts : 2D array_like of ints or floats
+        Matrix containing count/abundance data where each row contains counts
+        of OTUs in a given sample.
+    ids : iterable of strs
+        Identifiers for each sample in ``counts``.
+    validate : bool, optional
+        See ``skbio.diversity.beta_diversity`` for details.
+    reduce_f : function, optional
```

Comment:
I think I'm starting to understand, this is the "reduce" step of map-reduce. Maybe point that out in the docs? I assume this exists so that results can be merged piecewise.